relay-threading/src/pool.rs

```diff
@@ -0,0 +1,371 @@
+use std::any::Any;
+use std::future::Future;
+use std::io;
+use std::panic::AssertUnwindSafe;
+use std::sync::Arc;
+
+use futures::future::BoxFuture;
+use futures::FutureExt;
+
+use crate::builder::AsyncPoolBuilder;
+use crate::multiplexing::Multiplexed;
+
+/// [`AsyncPool`] is a thread-based executor that runs asynchronous tasks on dedicated worker threads.
+///
+/// The pool collects tasks through a bounded channel and distributes them among threads, each of which runs its own
+/// Tokio executor. This design enables controlled concurrency and efficient use of system resources.
+#[derive(Debug)]
+pub struct AsyncPool<F> {
+    tx: flume::Sender<F>,
+}
+
+impl<F> AsyncPool<F>
+where
+    F: Future<Output = ()> + Send + 'static,
+{
+    /// Creates a new [`AsyncPool`] based on the configuration specified by [`AsyncPoolBuilder`].
+    ///
+    /// This method initializes the dedicated worker threads and configures each executor with the defined
+    /// concurrency limits.
+    pub fn new<S>(mut builder: AsyncPoolBuilder<S>) -> io::Result<Self>
+    where
+        S: ThreadSpawn,
+    {
+        let (tx, rx) = flume::bounded(builder.num_threads * 2);
+
+        for index in 0..builder.num_threads {
+            let rx = rx.clone();
+
+            let thread = Thread {
+                index,
+                max_concurrency: builder.max_concurrency,
+                name: builder.thread_name.as_mut().map(|f| f(index)),
+                runtime: builder.runtime.clone(),
+                panic_handler: builder.thread_panic_handler.clone(),
+                task: Multiplexed::new(
+                    builder.max_concurrency,
+                    rx.into_stream(),
+                    builder.task_panic_handler.clone(),
+                )
+                .boxed(),
+            };
+
+            builder.spawn_handler.spawn(thread)?;
+        }
+
+        Ok(Self { tx })
+    }
+}
+
+impl<F> AsyncPool<F>
+where
+    F: Future<Output = ()>,
+{
+    /// Schedules a future for execution within the [`AsyncPool`].
+    ///
+    /// The task is added to the pool's internal queue to be executed by an available worker thread.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if the internal task channel is unexpectedly closed.
+    pub fn spawn(&self, future: F) {
+        assert!(
+            self.tx.send(future).is_ok(),
+            "receiver never exits before sender"
```

Comment:
Can come up with a better error here, this can actually happen now.

And also the docs should reflect this.